src/_zkapauthorizer/_storage_server.py

```diff
@@ -695,3 +701,100 @@ def get_stat(sharepath):
             return stat_slot
         else:
             return stat_bucket
+
+
+def add_leases_for_writev(storage_server, storage_index, secrets, tw_vectors, now):
+    """
+    Add a new lease using the given secrets to all shares written by
+    ``tw_vectors``.
+    """
+    for (sharenum, sharepath) in get_all_share_paths(storage_server, storage_index):
+        testv, datav, new_length = tw_vectors.get(sharenum, (None, b"", None))
+        if datav or (new_length is not None):
+            # It has data or a new length - it is a write.
+            if share_has_active_leases(storage_server, storage_index, sharenum, now):
+                # It's fine, leave it be.
+                continue
+
+            # Aha.  It has no lease that hasn't expired.  Give it one.
+            (write_enabler, renew_secret, cancel_secret) = secrets
+            share = get_share_file(sharepath)
+            share.add_or_renew_lease(
+                LeaseInfo(
+                    owner_num=1,
+                    renew_secret=renew_secret,
+                    cancel_secret=cancel_secret,
+                    expiration_time=now
+                    + ZKAPAuthorizerStorageServer.LEASE_PERIOD.total_seconds(),
+                    nodeid=storage_server.my_nodeid,
+                ),
+            )
+
+
+def get_share_path(storage_server, storage_index, sharenum):
+    # type: (StorageServer, bytes, int) -> FilePath
+    """
+    Get the path to the given storage server's storage for the given share.
+    """
+    return (
+        FilePath(storage_server.sharedir)
+        .preauthChild(storage_index_to_dir(storage_index))
+        .child(u"{}".format(sharenum))
+    )
+
+
+def share_has_active_leases(storage_server, storage_index, sharenum, now):
+    # type: (StorageServer, bytes, int, float) -> bool
+    """
+    Determine whether the given share on the given server has an unexpired
+    lease or not.
+
+    :return: ``True`` if it has at least one unexpired lease, ``False``
+        otherwise.
+    """
+    sharepath = get_share_path(storage_server, storage_index, sharenum)
+    share = get_share_file(sharepath.path)
+    return any(lease.get_expiration_time() > now for lease in share.get_leases())
```

Comment:
I *think* this should probably have an `if lease.renew_secret == renew_secret` here, but that will require changes elsewhere.

Reply:
It's possible.  I think it will depend on how we want billing and sharing to interact, though.  Since we're basically just not doing sharing yet I think any interaction (eg, the current behavior) is fine for now.